### DISCUSSION TITLE
Pass data from middleware to controller

### DISCUSSION THREAD

**Discussion by unkobweb** [adonisjs/core#2877] [Help] [answered]
Hi,

I use Adonis for a microservices application, my "documents" microservice get a middleware who send the Bearer token to the auth middleware to know if the token is valid and at which user it is attached. 

The auth middleware send back the id of the allowed user and after i need to pass this user id to my controller function (create a document and set the id of the owner) for the moment i did something with the session but i think it's dangerous because i guess there is just one session (between middlewares) and im pretty sure if two request are really close the user id can be override and cause problems to my application

There is the auth middleware : 
```js
const auth_path = Env.get('auth_path', '127.0.0.1:3000')
const token = request.header('authorization')
if (!token) {
  response.unauthorized("E_INVALID_API_TOKEN: Invalid API token")
  return
}
await axios.get(`http://${auth_path}/api/v1/authenticate`, {
  headers: {
    Authorization: token
  }
}).then(async (res) => {
  session.put('user', res.data)
  await next()
}).catch(() => {
  response.unauthorized("E_INVALID_API_TOKEN: Invalid API token")
})
```

And i get the user id in my controller like that : 
```js
const user = session.get('user')
```

So my questions are : 
- There is a way to pass data from the middleware to the controller (write it in the context ?)
- My way with session is it dangerous ? (really just one session between microservices or not)

My other solution is just to put my middleware code in a controller function and call it at the start of all my other controller functions

Sorry for my bad english, i try to do my best ^^

Best regards,
Kobweb

**Comment by arthurcvm** [comment, marked as answer]
Hi @unkobweb, you can manipulate the request object in middleware, look this example:

```ts
import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'

export default class IndexMiddleware {
  public async handle ({ request }: HttpContextContract, next: () => Promise<void>): Promise<void> {
    let filter = {}
    try {
      filter = request.get().filter ? JSON.parse(request.get().filter) : {}
    } catch (error) {
      console.warn('FilterMiddleware: ', error)
    }
    request.all().page ||= 1
    request.all().per_page ||= 10
    request.all().order_column ||= null
    request.all().order ||= null
    request.all().filter = filter

    await next()
  }
}
```
To pass data to controller from the auth middleware you can deconstruct the `auth` object in `HttpContextContract`, look this:
```ts
public async index ({ response, request, auth }: HttpContextContract): Promise<void> {
    const user = auth.user as User
    ///continue...
}
```
